docs/architecture/adr-004-genesis-modes.md

```diff
@@ -0,0 +1,253 @@
+# ADR 004: Multiple Genesis modes
+
+## Changelog
+
+- Sept 3, 2019: Initial draft
+
+## Context
+
+In the SDK, when the `GenesisState` is initialized through a `InitGenesis` call, some of the `GenesisState` fields are loaded from a `genesis.json` file, whereas in other cases they are initialized (_i.e_ populated) if the corresponding object value from JSON is empty, undefined or its value is zero.
+
+This may lead to confusion because as is currently programmed, `InitGenesis`
+initializes _and_ loads the values at the same time. When in reality they should be
+two separated disctinct processes.
+
+A recurring case of this problem is to check if a balance is empty and initialize it
+if not (by iterating the state). This affects mostly the total [`Supply`](https://github.com/cosmos/cosmos-sdk/blob/13e5e18d77e6010a4566ce187f18207669345419/x/supply/genesis.go#L14-L24) and[`ModuleAccounts`](https://github.com/cosmos/cosmos-sdk/blob/13e5e18d77e6010a4566ce187f18207669345419/x/gov/genesis.go#L44-L50)
+within each module.
+
+## Decision
+
+We propose to add a separate command to update a genesis file with the missing values/fields for each module `GenesisState`.
+
+With this command, the user provides a `genesis.json` file with the minimum amount
+of genesis fields required to start a chain. The rest is loaded directly from the
+mandatory genesis parameters. In general, the genesis fields that are populated
+are the ones checked by the simulation invariants.
+
+If the given genesis file contains _any_ field that shouldn't be populated, the
+execution should `panic` and exit. Similarly, the initialization of the app (_i.e_
+`<app>d init`) should `panic` if any genesis field is undefined.
```

Comment:
I agree with this simplicity in design. The blockchain developer obviously needs to have the capacity to modify what populatable fields are and by which logic they are populated.